File: examples/threads/ProcessExample.java
import java.io.IOException;

/**
 * Demonstrates how to create child processes and wait for them to finish. Run
 * the program and then from a console check the processes created using the
 * following command:
 * 
 * ps augx | grep sleep
 * 
 * @author amit
 */
public class ProcessExample
{
    private static final int NUM_PROCESS = 5;

    public static void main(String[] args) throws IOException, InterruptedException {
        Process[] child = new Process[NUM_PROCESS];
        for (int i = 0; i < NUM_PROCESS; i++) {
            child[i] = new ProcessBuilder("sleep", "30").start();
            System.out.println("Child process " + i + " created");
        }
        System.out.println();

        for (int i = 0; i < NUM_PROCESS; i++) {
            child[i].waitFor();
            System.out.println("Child process " + i + " finished");
        }
        System.out.println();
    }
}

File: examples/other-languages/go/threads/ProcessExample.go
//Author: Amit Jain & Paul Ellis
package main

import (
	"fmt"
	"os/exec"
	"strconv"
)

const MAX_PROCESSES = 5

func main(){

	//Create a slice to hold 5 processes
	procs := make([]*exec.Cmd, MAX_PROCESSES)
	
	//create and start a sleep command for each space in the slice
	//and save to the slice
	for i := 0; i < MAX_PROCESSES; i++{
		procs[i] = exec.Command("sleep", "30")
		procs[i].Start()
		fmt.Println("Child Process " + strconv.Itoa(i) + " created")
	}
	fmt.Println()

	//wait for each command in the slice to finish
	for i := 0; i < MAX_PROCESSES; i++{
		procs[i].Wait()
		fmt.Println("Child Process " + strconv.Itoa(i) + " finished")
	}

	fmt.Println()
}

